### PR TITLE
Object Storage Integration

### PR DESCRIPTION
**Pull request by lepsalex**
This is pretty significant refactor, the goals were the following:
- componentize reusable functionality so that we can compose different archive build pipelines more efficiently (https://wiki.oicr.on.ca/display/OV/CanCoGen+Virus+Seq+Data+Portal#CanCoGenVirusSeqDataPortal-4.4.1DataDownloadandAggregation-HighLevel)
- drop the usage of Muse in favour of going straight object storage using the AWS SDK
- update the API to allow for downloading files from object storage based on `archiveId` or the latest "all" archive
- rework the all archive build process to immediately cancel when new events start coming in if mid-build
- have the archive build process check that it is not building a duplicate archive, more relevant for future set query but also very useful if we wanted to use the interval builder instead of the event-driven one

**NOTE: The app config has changed, we will need to update the deployment values**

### CODE REVIEW

File: src/main/java/org/cancogenvirusseq/singularity/components/InstantToArchiveBuildRequest.java [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+    return Mono.just(
+            new SearchSourceBuilder()
+                .query(QueryBuilders.rangeQuery(AnalysisDocument.LAST_UPDATED_AT_FIELD).to(instant))
+                .aggregation(
+                    AggregationBuilders.max("lastUpdatedDate")
+                        .field(AnalysisDocument.LAST_UPDATED_AT_FIELD))
+                .aggregation(AggregationBuilders.count("totalHits").field("_id"))
+                .fetchSource(false))
+        .map(
+            source ->
```
Comment (andricDu) [inline src/main/java/org/cancogenvirusseq/singularity/components/InstantToArchiveBuildRequest.java]: why not used named methods here as references. That way the whole chain would read like prose.

File: src/main/java/org/cancogenvirusseq/singularity/components/hoc/InstantToArchiveBuildRequest.java [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+  private final ReactiveElasticSearchClientConfig reactiveElasticSearchClientConfig;
+  private final ArchivesRepo archivesRepo;
+
+  @Override
+  public Mono<ArchiveBuildRequest> apply(Instant instant) {
+    return Mono.just(
+            new SearchSourceBuilder()
+                .query(QueryBuilders.rangeQuery(AnalysisDocument.LAST_UPDATED_AT_FIELD).to(instant))
+                .aggregation(
+                    AggregationBuilders.max("lastUpdatedDate")
```
Comment (andricDu) [inline src/main/java/org/cancogenvirusseq/singularity/components/hoc/InstantToArchiveBuildRequest.java:40]: why not use a static constant for these names?

File: src/main/java/org/cancogenvirusseq/singularity/config/s3Client/S3ClientProperties.java [hunk 3]
```diff
@@ -16,36 +16,28 @@
...
+  private String secretAccessKey;
+
+  // Bucket name we'll be using as our backend storage
+  private String bucket;
+  private String dataDir;
+
+  // AWS S3 requires that file parts must have at least 5MB, except
+  // for the last part. This may change for other S3-compatible services, so let't
+  // define a configuration property for that
+  private int multipartMinPartSize = 5 * 1024 * 1024;
```
Comment (andricDu) [inline src/main/java/org/cancogenvirusseq/singularity/config/s3Client/S3ClientProperties.java:42]: As far as I can tell, you are not actually doing any multipart uploads?

File: src/main/java/org/cancogenvirusseq/singularity/repository/model/Archive.java [hunk 4]
```diff
@@ -22,7 +19,7 @@
   private Long createdAt;
 
   @NonNull private Integer numOfSamples;
-  @NonNull private Integer numOfDownloads;
+  private Integer numOfDownloads;
```
Comment (andricDu) [inline src/main/java/org/cancogenvirusseq/singularity/repository/model/Archive.java:24]: why is this nullable?

**Review by andricDu — CHANGES_REQUESTED** [review]
Looks good, some minor stuff. 

Really don't like long chains of lambdas as it forces the reader to focus on the nitty gritty rather than the overall functional intent. Much prefer description method reference chaining.

File: src/main/java/org/cancogenvirusseq/singularity/repository/model/Archive.java (hunk 4 above)
Comment (lepsalex) [reply src/main/java/org/cancogenvirusseq/singularity/repository/model/Archive.java:24]: it always starts of at 0 which is the default

**Review by jaserud — APPROVED** [review]

**Pull request merged**